# Panda Notes

# Series (Series in Panda now instead of Numpy)


In [1]:
import numpy as np 
import pandas as pd

Series are similar to NumpYx arrays, except that we can give then a named or datetime index, instead of just a numerical index

Series is later on used to work with data frames

In [2]:
labels = ['a','b','c']
list = [10,20,30]

arr = np.array([10,20,30])


The above are the starting points to creating a list.

The items needed are the data argument, and the index argument. This is what differentiates the series from a normal NumPy array

The below illustrates

In [3]:
labels = ['a','b','c']
list = [10,20,30]

pd.Series (list, index= labels)

a    10
b    20
c    30
dtype: int64

A quick note is that the index = does not have to be specified for the .Series command to work

as seen below

In [4]:
arr = np.array([10,20,30])
labels = ['a','b','c']

pd.Series (arr, labels)

a    10
b    20
c    30
dtype: int32

The following example shows that the dictionary can also be used in the .Series command and the labels, when matching to the keys of the dictionary, will allow the values to be extracted into the series format  

In [5]:
dict = {'a':10,'b':10,'c':10}
labels = ['a','b','c']

pd.Series (dict, labels)

a    10
b    10
c    10
dtype: int64

Accessing values within the series

In [6]:
Ser1 = pd.Series ([1,2,3,4],index = ['USA','CHINA','FRANCE','GERMANY'])

Ser1

USA        1
CHINA      2
FRANCE     3
GERMANY    4
dtype: int64

In [7]:
Ser1 = pd.Series ([1,2,3,4],index = ['USA','CHINA','FRANCE','GERMANY'])

Ser1 ['USA']

1

In [8]:
Ser2 = pd.Series ([10,20,30,40],index = ['SG','CHINA','FRANCE','ITALY'])

Ser2 

SG        10
CHINA     20
FRANCE    30
ITALY     40
dtype: int64

The following Ser1 + Ser2 is to show that the addition or any other math operator on the value only works if the index if matching across the two series 

In [9]:
Ser1 + Ser2


CHINA      22.0
FRANCE     33.0
GERMANY     NaN
ITALY       NaN
SG          NaN
USA         NaN
dtype: float64

In [10]:
Ser1 * Ser2

CHINA      40.0
FRANCE     90.0
GERMANY     NaN
ITALY       NaN
SG          NaN
USA         NaN
dtype: float64

## --------------------------

## Dataframes 

In [20]:
from numpy.random import randn

In [21]:
np.random.seed(101)

Do not be concerned with np.random.seed, it is just a command to ensure that the random values genereated in this personal set of notes is consistent with that of the instructor's

## Creating a dataframe

The syntax to create a dataframe is the following:

dataframe = pd.Dataframe( data , index , column headers) 

In [22]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

Notice here that the 2 lists: ['A','B','C','D','E'] and ['W','X','Y','Z'] are
again in the conventional row, column matrix order

Also note that the (5,4) in the randn function simply refers to the number of rows and columns of the array values

In [26]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']

df

A = np.mean(df ['New'])
print(df)

print('blank')

print(A)

          W         X         Y         Z       New
A -1.467514 -0.494095 -0.162535  0.485809 -1.961609
B  0.392489  0.221491 -0.855196  1.541990  0.613979
C  0.666319 -0.538235 -0.568581  1.407338  0.128085
D  0.641806 -0.905100 -0.391157  1.028293 -0.263294
E -1.972605 -0.866885  0.720788 -1.223082 -2.839490
blank
-0.8644659432945913


The above example was included to show the interaction of the np.mean operation with the dataframe's newly created column

For another way, it is also possibel to create a dataframe out of a dictionary 

In [30]:
d ={'A':[1,2,3],'B':[5,6,7],'C':["A","B","C"]}

df = pd.DataFrame(d) 

df


A  B  C
0  1  5  A
1  2  6  B
2  3  7  C

## Accessing coloumns within the dataframes
## also know as Column SubSetting

In [15]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])
df

W         X         Y         Z
A -0.993263  0.196800 -1.136645  0.000366
B  1.025984 -0.156598 -0.031579  0.649826
C  2.154846 -0.610259 -0.755325 -0.346419
D  0.147027 -0.479448  0.558769  1.024810
E -0.925874  1.862864 -1.133817  0.610478

In [16]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['W']

A    0.386030
B    0.681209
C   -1.005187
D   -1.382920
E    0.992573
Name: W, dtype: float64

Note that the above method does not work for grabbing 'sideways' 

Eg. df ['A'] results in an error


For subsetting multiple columns, remember to put the mutiple columns within a list, or either that a list viarble can be defined 
separately and later called on through dataframe [list]

In [17]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df[['W','Y']]

W         Y
A -1.467514 -0.162535
B  0.392489 -0.855196
C  0.666319 -0.568581
D  0.641806 -0.391157
E -1.972605  0.720788

In [18]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

filter = ['W','Z']

df[filter]

W         Z
A  1.606780 -1.329660
B  0.041460  0.110477
C -0.804652  0.358941
D -1.080471  0.833029
E  0.975720 -0.708954

The above example shows that multiple columns can be requested at the same time


## The following dataset is included to show subsetting by specific value of the column (category)
## aka filtering data based on a specific category


In [19]:
cars = pd.read_csv(r"C:\Users\lauko\OneDrive\Desktop\DataCamp\Automobile_data.csv") 
cars.head()

index      company   body-style  wheel-base  length engine-type  \
0      0  alfa-romero  convertible        88.6   168.8        dohc   
1      1  alfa-romero  convertible        88.6   168.8        dohc   
2      2  alfa-romero    hatchback        94.5   171.2        ohcv   
3      3         audi        sedan        99.8   176.6         ohc   
4      4         audi        sedan        99.4   176.6         ohc   

  num-of-cylinders  horsepower  average-mileage    price  
0             four         111               21  13495.0  
1             four         111               21  16500.0  
2              six         154               19  16500.0  
3             four         102               24  13950.0  
4             five         115               18  17450.0

In [20]:
sedan_cars = cars[cars['body-style'] == 'sedan']
sedan_cars.head()

index company body-style  wheel-base  length engine-type num-of-cylinders  \
3      3    audi      sedan        99.8   176.6         ohc             four   
4      4    audi      sedan        99.4   176.6         ohc             five   
5      5    audi      sedan        99.8   177.3         ohc             five   
7      9     bmw      sedan       101.2   176.8         ohc             four   
8     10     bmw      sedan       101.2   176.8         ohc             four   

   horsepower  average-mileage    price  
3         102               24  13950.0  
4         115               18  17450.0  
5         110               19  15250.0  
7         101               23  16430.0  
8         101               23  16925.0

Notice that the cars['body-style'] within cars[cars['body-style'] == 'sedan'] is referring to the particular column

In [21]:
bigwheel_cars = cars[cars['wheel-base'] >100]
bigwheel_cars.head()

index company body-style  wheel-base  length engine-type num-of-cylinders  \
6       6    audi      wagon       105.8   192.7         ohc             five   
7       9     bmw      sedan       101.2   176.8         ohc             four   
8      10     bmw      sedan       101.2   176.8         ohc             four   
9      11     bmw      sedan       101.2   176.8         ohc              six   
10     13     bmw      sedan       103.5   189.0         ohc              six   

    horsepower  average-mileage    price  
6          110               19  18920.0  
7          101               23  16430.0  
8          101               23  16925.0  
9          121               21  20970.0  
10         182               16  30760.0

In [22]:
body = cars['body-style'] == 'sedan'
bigwheel = cars['wheel-base'] >100

cars[body & bigwheel]

index        company body-style  wheel-base  length engine-type  \
7       9            bmw      sedan       101.2   176.8         ohc   
8      10            bmw      sedan       101.2   176.8         ohc   
9      11            bmw      sedan       101.2   176.8         ohc   
10     13            bmw      sedan       103.5   189.0         ohc   
11     14            bmw      sedan       103.5   193.8         ohc   
12     15            bmw      sedan       110.0   197.0         ohc   
24     33         jaguar      sedan       113.0   199.6        dohc   
25     34         jaguar      sedan       113.0   199.6        dohc   
26     35         jaguar      sedan       102.0   191.7        ohcv   
31     43          mazda      sedan       104.9   175.0         ohc   
32     44  mercedes-benz      sedan       110.0   190.9         ohc   
34     46  mercedes-benz      sedan       120.9   208.1        ohcv   
44     57         nissan      sedan       100.4   184.6        ohcv   
59     87          volvo      sedan       104.3   188.8         ohc   

   num-of-cylinders  horsepower  average-mileage    price  
7              four         101               23  16430.0  
8              four         101               23  16925.0  
9               six         121               21  20970.0  
10              six         182               16  30760.0  
11              six         182               16  41315.0  
12              six         182               15  36880.0  
24              six         176               15  32250.0  
25              six         176               15  35550.0  
26           twelve         262               13  36000.0  
31             four          72               31  18344.0  
32             five         123               22  25552.0  
34            eight         184               14  40960.0  
44              six         152               19  13499.0  
59             four         114               23  12940.0

## Subsetting using .isin () method 
## aka picking out several particular values within a given category / list and using that as filter 

In [23]:
cars = pd.read_csv(r"C:\Users\lauko\OneDrive\Desktop\DataCamp\Automobile_data.csv") 
cars.head()

index      company   body-style  wheel-base  length engine-type  \
0      0  alfa-romero  convertible        88.6   168.8        dohc   
1      1  alfa-romero  convertible        88.6   168.8        dohc   
2      2  alfa-romero    hatchback        94.5   171.2        ohcv   
3      3         audi        sedan        99.8   176.6         ohc   
4      4         audi        sedan        99.4   176.6         ohc   

  num-of-cylinders  horsepower  average-mileage    price  
0             four         111               21  13495.0  
1             four         111               21  16500.0  
2              six         154               19  16500.0  
3             four         102               24  13950.0  
4             five         115               18  17450.0

In [24]:
powerful_cylinders = cars['num-of-cylinders'].isin(['five','six'])

cars [powerful_cylinders].head()

index      company body-style  wheel-base  length engine-type  \
2      2  alfa-romero  hatchback        94.5   171.2        ohcv   
4      4         audi      sedan        99.4   176.6         ohc   
5      5         audi      sedan        99.8   177.3         ohc   
6      6         audi      wagon       105.8   192.7         ohc   
9     11          bmw      sedan       101.2   176.8         ohc   

  num-of-cylinders  horsepower  average-mileage    price  
2              six         154               19  16500.0  
4             five         115               18  17450.0  
5             five         110               19  15250.0  
6             five         110               19  18920.0  
9              six         121               21  20970.0

## The following is an additional note on how to sort an example table like the one above in ascending or decesending order.

## This is done using the .sort_values(column, ascending = True/False) function 

In [25]:
POWERFUL_Cylinders = cars['num-of-cylinders'].isin(['five','six'])

powerful_cylinder_cars = cars [POWERFUL_Cylinders]

powerful_cylinder_cars.sort_values('num-of-cylinders', ascending = False) 

index        company   body-style  wheel-base  length engine-type  \
2       2    alfa-romero    hatchback        94.5   171.2        ohcv   
9      11            bmw        sedan       101.2   176.8         ohc   
10     13            bmw        sedan       103.5   189.0         ohc   
11     14            bmw        sedan       103.5   193.8         ohc   
12     15            bmw        sedan       110.0   197.0         ohc   
24     33         jaguar        sedan       113.0   199.6        dohc   
25     34         jaguar        sedan       113.0   199.6        dohc   
44     57         nissan        sedan       100.4   184.6        ohcv   
45     61        porsche      hardtop        89.5   168.9        ohcf   
46     62        porsche  convertible        89.5   168.9        ohcf   
54     79         toyota        wagon       104.5   187.8        dohc   
4       4           audi        sedan        99.4   176.6         ohc   
5       5           audi        sedan        99.8   177.3         ohc   
6       6           audi        wagon       105.8   192.7         ohc   
32     44  mercedes-benz        sedan       110.0   190.9         ohc   
33     45  mercedes-benz        wagon       110.0   190.9         ohc   

   num-of-cylinders  horsepower  average-mileage    price  
2               six         154               19  16500.0  
9               six         121               21  20970.0  
10              six         182               16  30760.0  
11              six         182               16  41315.0  
12              six         182               15  36880.0  
24              six         176               15  32250.0  
25              six         176               15  35550.0  
44              six         152               19  13499.0  
45              six         207               17  34028.0  
46              six         207               17  37028.0  
54              six         156               19  15750.0  
4              five         115               18  17450.0  
5              five         110               19  15250.0  
6              five         110               19  18920.0  
32             five         123               22  25552.0  
33             five         123               22  28248.0

.sort_values ( ) can also be used for multi variables

In [26]:
cars = pd.read_csv(r"C:\Users\lauko\OneDrive\Desktop\DataCamp\Automobile_data.csv") 
cars.sort_values (["horsepower","price"],ascending = [True, False]) 

index        company   body-style  wheel-base  length engine-type  \
13     16      chevrolet    hatchback        88.4   141.1           l   
57     82     volkswagen        sedan        97.3   171.7         ohc   
55     80     volkswagen        sedan        97.3   171.7         ohc   
40     53         nissan        sedan        94.5   165.3         ohc   
53     71         toyota        wagon        95.7   169.7         ohc   
..    ...            ...          ...         ...     ...         ...   
34     46  mercedes-benz        sedan       120.9   208.1        ohcv   
46     62        porsche  convertible        89.5   168.9        ohcf   
45     61        porsche      hardtop        89.5   168.9        ohcf   
26     35         jaguar        sedan       102.0   191.7        ohcv   
47     63        porsche    hatchback        98.4   175.7       dohcv   

   num-of-cylinders  horsepower  average-mileage    price  
13            three          48               47   5151.0  
57             four          52               37   7995.0  
55             four          52               37   7775.0  
40             four          55               45   7099.0  
53             four          62               27   8778.0  
..              ...         ...              ...      ...  
34            eight         184               14  40960.0  
46              six         207               17  37028.0  
45              six         207               17  34028.0  
26           twelve         262               13  36000.0  
47            eight         288               17      NaN  

[61 rows x 10 columns]

## Accessing rows within the dataframes (use of method is involved) 

## also know as Row SubSetting

In [27]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']

print (df)

df.loc['A']

          W         X         Y         Z       New
A  0.586847 -1.621348  0.677535  0.026105 -1.034501
B -1.678284  0.333973 -0.532471  2.117727 -1.344311
C  0.197524  2.302987  0.729024 -0.863091  2.500511
D  0.305632  0.243178  0.864165 -1.560931  0.548809
E -0.251897 -0.578120  0.236996  0.200780 -0.830017


W      0.586847
X     -1.621348
Y      0.677535
Z      0.026105
New   -1.034501
Name: A, dtype: float64

The above also leads to the concept that rows, along with columns are also a series of values

The below shows the 2nd way of obtaining data in a row

In [28]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']

print(df)

df.iloc [2]

          W         X         Y         Z       New
A  0.327845  0.674485 -0.174057  0.780140  1.002330
B -0.383258 -0.409318  0.343539  0.196275 -0.792577
C -0.982776  2.231555 -0.971393 -1.522333  1.248779
D  1.133703  0.528187  0.393461 -0.630507  1.661890
E -1.398290 -0.219311 -0.045676  0.012421 -1.617600


W     -0.982776
X      2.231555
Y     -0.971393
Z     -1.522333
New    1.248779
Name: C, dtype: float64

The iloc method is different from the loc method in that it uses the index numerical number, while the previous loc uses the index label 

## Accessing specific value/subset within the dataframes 

Specific Value within the dataframe

In [29]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']

df.loc['B','Y']

-2.7369945956467303

Subset of Values within the dataframe

In [30]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']

df.loc [['A','B'], ['W','Y']]

W         Y
A -1.306482  0.802129
B  0.608857 -0.205792

## Creating/Dropping Columns/Rows within the dataframe

The below following examples are to illustrate the creating of new columns

In [31]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']

df

W         X         Y         Z       New
A -0.279510  1.062710  1.752014  0.695547  0.783200
B  0.153661  0.167638 -0.765930  0.962299  0.321299
C  0.902826 -0.537909 -1.549671  0.435253  0.364917
D  1.259904 -0.447898  0.266207  0.412580  0.812005
E  0.988773  0.513833 -0.928205  0.846904  1.502606

The below following examples are to illustrate the dropping/deletion of columns using the exsisting df dataframe that has been stored in memory 

In [32]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']


df.drop('New',axis=1, inplace=True)
df

W         X         Y         Z
A -0.298436  0.029141  0.889031 -1.839261
B  0.863596 -1.076591 -0.580383  1.209027
C  1.201744 -0.820463 -1.029577 -1.524856
D -0.854860  1.602816  0.185479 -0.994018
E -0.131906 -1.335092 -0.089298  0.698566

For the df.drop syntax, notice the use of the inplace = True as an argument, and the purpose here is to confirm whether the dataframe wants to be overwritten when dropping/deleting columns


df.drop can also be also used to drop rows, by using the axis = 0

In [33]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

df ['New'] = df ['W'] + df ['X']


df.drop('E', axis=0, inplace=True)
df

W         X         Y         Z       New
A -1.409126  0.870907  1.585812  0.929658 -0.538219
B -0.550876  0.942045 -0.975349 -1.054851  0.391169
C  2.165421 -1.267240 -0.802843 -0.253479  0.898181
D  0.090699  1.321715  0.794129  1.359712  1.412414

## Conditional Selection in Dataframes

In [34]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

In [35]:
df [df > 0]

W         X         Y         Z
A       NaN       NaN       NaN  2.090539
B  1.004056  1.031916       NaN       NaN
C       NaN  0.292795  0.834355       NaN
D  0.119145  0.665715  0.407586  0.807843
E       NaN       NaN       NaN       NaN

The above shows the use of the whole frame being > 0 as a condition

This will not be very common. Instead, what is common is the use of single columns or rows as conditions. 

Example below illustrates. Frame only shows values where W column cells have a positive value

In [36]:
df [df['W'] > 0]

W         X         Y         Z
B  1.004056  1.031916 -1.081118 -0.000355
D  0.119145  0.665715  0.407586  0.807843

In [37]:
A = df [df['W'] > 0]

A [A['Z'] > 0]

W         X         Y         Z
D  0.119145  0.665715  0.407586  0.807843

The use of A in the above example to hold the new list after the first condition,

is to set up for the second condition 

An alternative would be to stack commands, as illustrated below:

In [38]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z']) 

print(df)

df [(df['W']> 0) & (df['Y'] < 0)]

          W         X         Y         Z
A -1.087436 -2.567767  0.661029 -0.332921
B -0.928602  1.715950 -0.468756  0.860097
C  0.230087  0.618658 -2.052563 -0.166646
D -0.445873  0.685874  2.785010  1.523750
E  1.290230 -1.478319  0.214234 -0.240510


W         X         Y         Z
C  0.230087  0.618658 -2.052563 -0.166646

Note: 

For the above example of multiple conditions, the following syntax: 

df [(df['W']> 0) and (df['Y'] < 1)]  WOULD NOT work, and this is because

the boolean values TRUE and FALSE will be confused (FALSE overrides TRUE) 

In [39]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z']) 

print(df)

df [(df['W']> 0) | (df['Y'] < 1)]

          W         X         Y         Z
A -1.420949  0.321660  1.007324  0.329784
B  0.920525  1.349138 -0.377861  1.955420
C -0.858304 -0.222210 -0.051203 -0.439128
D -1.501041  1.058357  1.854967  0.502250
E  0.114581 -0.368487  0.459343  0.585513


W         X         Y         Z
B  0.920525  1.349138 -0.377861  1.955420
C -0.858304 -0.222210 -0.051203 -0.439128
E  0.114581 -0.368487  0.459343  0.585513

The above example is to highlight the signage " | " 

Just as & was used for the and condition, " | " is used for the or condition

## Resettting/Relabelling index column of the dataframe

In [40]:
df = pd.DataFrame (randn(5,4),['A','B','C','D','E'],['W','X','Y','Z']) 
df.reset_index()

index         W         X         Y         Z
0     A -0.181904 -0.603235 -2.552660 -0.031927
1     B  1.407189  0.060971 -1.542393 -0.821110
2     C -1.599171 -2.068836  1.509747  0.174683
3     D  0.463428 -2.206442  0.544159  0.698296
4     E  1.252944 -1.048347 -1.281320  1.276731

Could use the index as a way of checking the current index values 

In [41]:
Countries = 'SG NY US CH KR'.split()
print (Countries)

['SG', 'NY', 'US', 'CH', 'KR']


Side note, notice that the .split() method is a quicker way of making a list data structure

In [42]:
Countries = 'SG NY US CH KR'.split()
print (Countries) 

new_states = Countries

df ['States'] = new_states 

df.set_index ('States')

['SG', 'NY', 'US', 'CH', 'KR']


W         X         Y         Z
States                                        
SG     -0.181904 -0.603235 -2.552660 -0.031927
NY      1.407189  0.060971 -1.542393 -0.821110
US     -1.599171 -2.068836  1.509747  0.174683
CH      0.463428 -2.206442  0.544159  0.698296
KR      1.252944 -1.048347 -1.281320  1.276731

## Handling Missing Values in Dataframes 

## 1.(dropping missing values)


The first step here will be to create a dictionary data structure, for the creating of a dataframe drawing from the dictionary 

Note that the dictionary is intentionally filled with missing data, as from the use of numpy np.nan 

In [43]:
d ={'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}

df = pd.DataFrame(d) 

df

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3

The .dropna method is illustrated below 

What this method does is that it drops any rows or columns with empty values (depending on the axis that is selected) 

Recall that the axis 0 is for rows, while the axis 1 is for columns 

In [44]:
d ={'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}

df = pd.DataFrame(d) 

df.dropna(axis=0)

A    B  C
0  1.0  5.0  1

In [45]:
d ={'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}

df = pd.DataFrame(d) 

df.dropna(axis=1)

C
0  1
1  2
2  3

In [46]:
d ={'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}

df = pd.DataFrame(d) 

print(df)

df.dropna(thresh=2)

     A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3


A    B  C
0  1.0  5.0  1
1  2.0  NaN  2

In [47]:
d ={'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}

df = pd.DataFrame(d) 

print(df)

df.dropna(thresh=3)

     A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3


A    B  C
0  1.0  5.0  1

Another argument, aside from axis, is the thresh threshold argument.
The threshold value refers to the number of values that mus be present within the row, for the row to be kept 

## 2. (filling in missing values)


Filling in missing values is done through the .fillna(value= ) method

In [48]:
d ={'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}
df = pd.DataFrame(d) 

df

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3

In [49]:
df['A'].fillna(value= 8)

0    1.0
1    2.0
2    8.0
Name: A, dtype: float64

In [50]:
df['A'].fillna(value= df['A'].mean())

0    1.0
1    2.0
2    1.5
Name: A, dtype: float64

## Dataframe statistics and GroupBy in DataFrames

 Group by allows the grouping of rows of data together (by catogery) and performing math/stat operations

In [51]:
data = {'Company':['GOOG','GOOG','MSFT','MSFT','FB','FB'],
       'Person':['Sam','Charlie','Amy','Vanessa','Carl','Sarah'],
       'Sales':[200,120,340,124,243,350]} 

df = pd.DataFrame(data)

df

Company   Person  Sales
0    GOOG      Sam    200
1    GOOG  Charlie    120
2    MSFT      Amy    340
3    MSFT  Vanessa    124
4      FB     Carl    243
5      FB    Sarah    350

The .gropuby() method will group the rows together based off a column name.
For example, grouping by the company name 

In [52]:
by_company = df.groupby("Company")

print(by_company)

Side note: The print does not output the new table. By intution, what follows is the operation.

For instance, grouping by the company names, By the mean values of their sales figures 


In [53]:
data = {'Company':['GOOG','GOOG','MSFT','MSFT','FB','FB'],
       'Person':['Sam','Charlie','Amy','Vanessa','Carl','Sarah'],
       'Sales':[200,120,340,124,243,350]} 

df = pd.DataFrame(data)

by_company = df.groupby("Company")
individual_company_sales = by_company.sum()

print(individual_company_sales)

         Sales
Company       
FB         593
GOOG       320
MSFT       464


In [54]:
data = {'Company':['GOOG','GOOG','MSFT','MSFT','FB','FB'],
       'Person':['Sam','Charlie','Amy','Vanessa','Carl','Sarah'],
       'Sales':[200,120,340,124,243,350]} 

df = pd.DataFrame(data)

individual_company_sale = df.groupby("Company").sum()

print(individual_company_sale)

         Sales
Company       
FB         593
GOOG       320
MSFT       464


The above 2 approaches is just to show that the df.groupby for grouping dataframe can be stored in variable, or written together with the math operation in the same line of code

In [55]:
mean_sales = by_company.mean()

mean_sales

Sales
Company       
FB       296.5
GOOG     160.0
MSFT     232.0

The following illustrates the other possible operations, along with .describe that shows stats


In [56]:
by_company.min()

Person  Sales
Company                
FB          Carl    243
GOOG     Charlie    120
MSFT         Amy    124

In [57]:
by_company.std()

Sales
Company            
FB        75.660426
GOOG      56.568542
MSFT     152.735065

In [58]:
by_company.max()

Person  Sales
Company                
FB         Sarah    350
GOOG         Sam    200
MSFT     Vanessa    340

In [59]:
by_company.count()

Person  Sales
Company               
FB            2      2
GOOG          2      2
MSFT          2      2

In [60]:
by_company.describe()

Sales                                                        
        count   mean         std    min     25%    50%     75%    max
Company                                                              
FB        2.0  296.5   75.660426  243.0  269.75  296.5  323.25  350.0
GOOG      2.0  160.0   56.568542  120.0  140.00  160.0  180.00  200.0
MSFT      2.0  232.0  152.735065  124.0  178.00  232.0  286.00  340.0

In [61]:
by_company.describe().transpose()

Company              FB        GOOG        MSFT
Sales count    2.000000    2.000000    2.000000
      mean   296.500000  160.000000  232.000000
      std     75.660426   56.568542  152.735065
      min    243.000000  120.000000  124.000000
      25%    269.750000  140.000000  178.000000
      50%    296.500000  160.000000  232.000000
      75%    323.250000  180.000000  286.000000
      max    350.000000  200.000000  340.000000

Side note: The side note add on funciton here is just to reposition the table of stats

## Dataframe statistics using the .agg() method 

There is the defining of functions involved. 

In [62]:
cars = pd.read_csv(r"C:\Users\lauko\OneDrive\Desktop\DataCamp\Automobile_data.csv") 
cars.head()

index      company   body-style  wheel-base  length engine-type  \
0      0  alfa-romero  convertible        88.6   168.8        dohc   
1      1  alfa-romero  convertible        88.6   168.8        dohc   
2      2  alfa-romero    hatchback        94.5   171.2        ohcv   
3      3         audi        sedan        99.8   176.6         ohc   
4      4         audi        sedan        99.4   176.6         ohc   

  num-of-cylinders  horsepower  average-mileage    price  
0             four         111               21  13495.0  
1             four         111               21  16500.0  
2              six         154               19  16500.0  
3             four         102               24  13950.0  
4             five         115               18  17450.0

In [63]:
def percentile30(column):
    return column.quantile(0.3)

X = cars['horsepower'].agg(percentile30)

print(X)
print(f'The 30th percentile for horsepower is {X}')

69.0
The 30th percentile for horsepower is 69.0


## TimeSeries Data in Pandas (Datetime Index handling)

The datetime library is part of the pandas built in date and time library

In [12]:
import numpy as np 
import pandas as pd

from datetime import datetime 

In [4]:
my_year = 2020
my_month = 5
my_day = 7
my_hour = 15
my_minute = 0
my_second = 0 

In [7]:
my_date = datetime(my_year,my_month,my_day)
my_date

datetime.datetime(2020, 5, 7, 0, 0)

In [8]:
type(my_date)

datetime.datetime

In [9]:
my_date.month

5

The following shows an example of putting two dates into a list, and creating a data frame with this date/time as an index (although this will not be commonly used) since the csv files would already have the datetime

Firstly, put the date/time values into a list and then create the dataframe using the pd.DateTimeIndex(first_two) 

In [14]:
first_two = [datetime(2020,4,30),datetime(2020,5,1)]

Index_of_Dataframe = pd.DatetimeIndex(first_two)

Index_of_Dataframe

DatetimeIndex(['2020-04-30', '2020-05-01'], dtype='datetime64[ns]', freq=None)

Recall:

The syntax to create a dataframe is the following:

dataframe = pd.Dataframe( data , index , column headers) 

In [15]:
dataframe = pd.DataFrame(np.random.rand(2,2), Index_of_Dataframe, ["A" ,"B" ])
dataframe

A         B
2020-04-30  0.358764  0.352755
2020-05-01  0.637104  0.234037

The following shows how to get information on the index (assuming that there are many dates running throughout the index) 

In [16]:
dataframe.index.argmax()

1

dataframe.index.argmax() method to show the index value of the last index entry 

in other words the total number of dates within the index (do not forget to +1 to this number since index starts from 0)

In [17]:
dataframe.index.max()

Timestamp('2020-05-01 00:00:00')

dataframe.index.max() to show the final/latest date of the index

## TimeSeries Data in Pandas (Datetime Resampling)

The concept of using time resampling is similar to that of groupby, in grouping data according to specified intervals of time. 

The purpose of using the timedate object for this grouping purpose, is that it is smarter or easier when handling periods such as business quarters or years. The follwing examples will illustrate

In [ ]:
import numpy as np
import pandas as pd
import mtaplotlib.pyplot as plt
%matplotlib inline

In [36]:
df = pd.read_csv(r"C:\Users\lauko\OneDrive\Desktop\walmart_stock.csv")
df

Date       Open       High        Low      Close    Volume  \
0     2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800   
1     2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300   
2     2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200   
3     2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400   
4     2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300   
...          ...        ...        ...        ...        ...       ...   
1253  2016-12-23  69.430000  69.750000  69.360001  69.540001   4803900   
1254  2016-12-27  69.300003  69.820000  69.250000  69.699997   4435700   
1255  2016-12-28  69.940002  70.000000  69.260002  69.309998   4875700   
1256  2016-12-29  69.209999  69.519997  69.120003  69.260002   4298400   
1257  2016-12-30  69.120003  69.430000  68.830002  69.120003   6889500   

      Adj Close  
0     52.619235  
1     52.078475  
2     51.825539  
3     51.459220  
4     51.616215  
...         ...  
1253  69.032411  
1254  69.191240  
1255  68.804087  
1256  68.754456  
1257  68.615479  

[1258 rows x 7 columns]

In [38]:
df['Date']=pd.to_datetime(df['Date'])

The above to_datetime ( ) syntax converts the string values of the dates into date objects (recall that objects are just instances of a class) 

In [39]:
df

Date       Open       High        Low      Close    Volume  \
0    2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800   
1    2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300   
2    2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200   
3    2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400   
4    2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300   
...         ...        ...        ...        ...        ...       ...   
1253 2016-12-23  69.430000  69.750000  69.360001  69.540001   4803900   
1254 2016-12-27  69.300003  69.820000  69.250000  69.699997   4435700   
1255 2016-12-28  69.940002  70.000000  69.260002  69.309998   4875700   
1256 2016-12-29  69.209999  69.519997  69.120003  69.260002   4298400   
1257 2016-12-30  69.120003  69.430000  68.830002  69.120003   6889500   

      Adj Close  
0     52.619235  
1     52.078475  
2     51.825539  
3     51.459220  
4     51.616215  
...         ...  
1253  69.032411  
1254  69.191240  
1255  68.804087  
1256  68.754456  
1257  68.615479  

[1258 rows x 7 columns]

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1258 entries, 0 to 1257
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       1258 non-null   datetime64[ns]
 1   Open       1258 non-null   float64       
 2   High       1258 non-null   float64       
 3   Low        1258 non-null   float64       
 4   Close      1258 non-null   float64       
 5   Volume     1258 non-null   int64         
 6   Adj Close  1258 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 68.9 KB


notice that the Dtype datatime for the date column is now datetime64

Recall now that setting the index is done through the .set_index (column, inplace = True) 

Do not forget the inplace, otherwise the index will not be changed

In [56]:
df = pd.read_csv(r"C:\Users\lauko\OneDrive\Desktop\walmart_stock.csv")
df['Date']=pd.to_datetime(df['Date'])

df.set_index('Date', inplace= True)

df

Open       High        Low      Close    Volume  Adj Close
Date                                                                       
2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800  52.619235
2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300  52.078475
2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200  51.825539
2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400  51.459220
2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300  51.616215
...               ...        ...        ...        ...       ...        ...
2016-12-23  69.430000  69.750000  69.360001  69.540001   4803900  69.032411
2016-12-27  69.300003  69.820000  69.250000  69.699997   4435700  69.191240
2016-12-28  69.940002  70.000000  69.260002  69.309998   4875700  68.804087
2016-12-29  69.209999  69.519997  69.120003  69.260002   4298400  68.754456
2016-12-30  69.120003  69.430000  68.830002  69.120003   6889500  68.615479

[1258 rows x 6 columns]

Now for the resampling, it is done through the syntax df.resample(rule = ' ') 

The following are the rules

<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="87%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Alias</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>B</td>
<td>business day frequency</td>
</tr>
<tr class="row-odd"><td>C</td>
<td>custom business day frequency (experimental)</td>
</tr>
<tr class="row-even"><td>D</td>
<td>calendar day frequency</td>
</tr>
<tr class="row-odd"><td>W</td>
<td>weekly frequency</td>
</tr>
<tr class="row-even"><td>M</td>
<td>month end frequency</td>
</tr>
<tr class="row-odd"><td>SM</td>
<td>semi-month end frequency (15th and end of month)</td>
</tr>
<tr class="row-even"><td>BM</td>
<td>business month end frequency</td>
</tr>
<tr class="row-odd"><td>CBM</td>
<td>custom business month end frequency</td>
</tr>
<tr class="row-even"><td>MS</td>
<td>month start frequency</td>
</tr>
<tr class="row-odd"><td>SMS</td>
<td>semi-month start frequency (1st and 15th)</td>
</tr>
<tr class="row-even"><td>BMS</td>
<td>business month start frequency</td>
</tr>
<tr class="row-odd"><td>CBMS</td>
<td>custom business month start frequency</td>
</tr>
<tr class="row-even"><td>Q</td>
<td>quarter end frequency</td>
</tr>
<tr class="row-odd"><td>BQ</td>
<td>business quarter endfrequency</td>
</tr>
<tr class="row-even"><td>QS</td>
<td>quarter start frequency</td>
</tr>
<tr class="row-odd"><td>BQS</td>
<td>business quarter start frequency</td>
</tr>
<tr class="row-even"><td>A</td>
<td>year end frequency</td>
</tr>
<tr class="row-odd"><td>BA</td>
<td>business year end frequency</td>
</tr>
<tr class="row-even"><td>AS</td>
<td>year start frequency</td>
</tr>
<tr class="row-odd"><td>BAS</td>
<td>business year start frequency</td>
</tr>
<tr class="row-even"><td>BH</td>
<td>business hour frequency</td>
</tr>
<tr class="row-odd"><td>H</td>
<td>hourly frequency</td>
</tr>
<tr class="row-even"><td>T, min</td>
<td>minutely frequency</td>
</tr>
<tr class="row-odd"><td>S</td>
<td>secondly frequency</td>
</tr>
<tr class="row-even"><td>L, ms</td>
<td>milliseconds</td>
</tr>
<tr class="row-odd"><td>U, us</td>
<td>microseconds</td>
</tr>
<tr class="row-even"><td>N</td>
<td>nanoseconds</td>
</tr>
</tbody>
</table>

In [60]:
df.sort_values('Date', ascending = True) 

Open       High        Low      Close    Volume  Adj Close
Date                                                                       
2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800  52.619235
2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300  52.078475
2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200  51.825539
2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400  51.459220
2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300  51.616215
...               ...        ...        ...        ...       ...        ...
2016-12-23  69.430000  69.750000  69.360001  69.540001   4803900  69.032411
2016-12-27  69.300003  69.820000  69.250000  69.699997   4435700  69.191240
2016-12-28  69.940002  70.000000  69.260002  69.309998   4875700  68.804087
2016-12-29  69.209999  69.519997  69.120003  69.260002   4298400  68.754456
2016-12-30  69.120003  69.430000  68.830002  69.120003   6889500  68.615479

[1258 rows x 6 columns]

In [47]:
df.head()

Date       Open       High        Low      Close    Volume  \
Date                                                                          
2012-01-03 2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800   
2012-01-04 2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300   
2012-01-05 2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200   
2012-01-06 2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400   
2012-01-09 2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300   

            Adj Close  
Date                   
2012-01-03  52.619235  
2012-01-04  52.078475  
2012-01-05  51.825539  
2012-01-06  51.459220  
2012-01-09  51.616215

In [49]:
mean_of_all_info_perYear = df.resample(rule='A').mean()
mean_of_all_info_perYear 

Open       High        Low      Close        Volume  \
Date                                                                   
2012-12-31  67.158680  67.602120  66.786520  67.215120  9.239015e+06   
2013-12-31  75.264048  75.729405  74.843055  75.320516  6.951496e+06   
2014-12-31  77.274524  77.740040  76.864405  77.327381  6.515612e+06   
2015-12-31  72.569405  73.064167  72.034802  72.491111  9.040769e+06   
2016-12-31  69.481349  70.019643  69.023492  69.547063  9.371645e+06   

            Adj Close  
Date                   
2012-12-31  59.389349  
2013-12-31  68.147179  
2014-12-31  71.709712  
2015-12-31  68.831426  
2016-12-31  68.054229

What the above table means across each row, is that the value is the mean value within the given year

Eg. from 2012-12-31 to 2013-12-31, the mean value of the Adj Close price was 68.147